js/web/test/unittests/backends/webgl/test-depth-to-space.ts

```diff
@@ -0,0 +1,177 @@
+      rawInput: new Float32Array([0., 1., 9., 10, 18., 19, 27., 28., 36., 37., 45., 46., 54., 55., 63., 64.]),
+      expectedOutput: new Float32Array([0, 18, 1, 19, 36, 54, 37, 55, 9, 27, 10, 28, 45, 63, 46, 64]),
+      mode: 'DCR',
+    },
+
+    {
+      elementCount: 48,
+      blocksize: 2,
+      inputShape: [1, 8, 2, 3],
+      outputShape: [1, 2, 4, 6],
+      inputTextureShape: [16, 3],
+      outputTextureShape: [8, 6],
+      rawInput: new Float32Array([
+        0.,  1.,  2.,  3.,  4.,  5.,  9.,  10., 11., 12., 13., 14., 18., 19., 20., 21.,
+        22., 23., 27., 28., 29., 30., 31., 32., 36., 37., 38., 39., 40., 41., 45., 46.,
+        47., 48., 49., 50., 54., 55., 56., 57., 58., 59., 63., 64., 65., 66., 67., 68.
+      ]),
+      expectedOutput: new Float32Array([
+        0.,  18., 1.,  19., 2.,  20., 36., 54., 37., 55., 38., 56., 3.,  21., 4.,  22.,
+        5.,  23., 39., 57., 40., 58., 41., 59., 9.,  27., 10., 28., 11., 29., 45., 63.,
+        46., 64., 47., 65., 12., 30., 13., 31., 14., 32., 48., 66., 49., 67., 50., 68.
+      ]),
+      mode: 'DCR',
+    },
+    {
+      elementCount: 8,
+      blocksize: 2,
+      inputShape: [1, 8, 1, 1],
+      outputShape: [1, 2, 2, 2],
+      inputTextureShape: [8, 1],
+      outputTextureShape: [4, 2],
+      rawInput: new Float32Array([0., 9., 18., 27., 36., 45., 54., 63.]),
+      expectedOutput: new Float32Array([0, 9, 18, 27, 36, 45, 54, 63]),
+      mode: 'CRD',
+    },
+    {
+      elementCount: 16,
+      blocksize: 2,
+      inputShape: [1, 8, 1, 2],
+      outputShape: [1, 2, 4, 2],
+      inputTextureShape: [1, 16],
+      outputTextureShape: [8, 2],
+      rawInput: new Float32Array([0., 1., 9., 10, 18., 19, 27., 28., 36., 37., 45., 46., 54., 55., 63., 64.]),
+      expectedOutput: new Float32Array([0, 9, 1, 10, 18, 27, 19, 28, 36, 45, 37, 46, 54, 63, 55, 64]),
+      mode: 'CRD',
+    },
+
+    {
+      elementCount: 48,
+      blocksize: 2,
+      inputShape: [1, 8, 2, 3],
+      outputShape: [1, 2, 4, 6],
+      inputTextureShape: [16, 3],
+      outputTextureShape: [8, 6],
+      rawInput: new Float32Array([
+        0.,  1.,  2.,  3.,  4.,  5.,  9.,  10., 11., 12., 13., 14., 18., 19., 20., 21.,
+        22., 23., 27., 28., 29., 30., 31., 32., 36., 37., 38., 39., 40., 41., 45., 46.,
+        47., 48., 49., 50., 54., 55., 56., 57., 58., 59., 63., 64., 65., 66., 67., 68.
+      ]),
+      expectedOutput: new Float32Array([
+        0.,  9.,  1.,  10., 2.,  11., 18., 27., 19., 28., 20., 29., 3.,  12., 4.,  13.,
+        5.,  14., 21., 30., 22., 31., 23., 32., 36., 45., 37., 46., 38., 47., 54., 63.,
+        55., 64., 56., 65., 39., 48., 40., 49., 41., 50., 57., 66., 58., 67., 59., 68.
+      ]),
+      mode: 'CRD',
+    },
+  ];
+}
+
+let backend: Backend|undefined;
+let sessionhandler: SessionHandler|undefined;
+let inferenceHandler: InferenceHandler|undefined;
+
+describe('#UnitTest# - unpacked WebGLDepthToSpace - Tensor WebGLDepthToSpace', () => {
```

Comment:
is it possible to use operator tests to test DepthToSpace and MatMul?

pack and unpack is special because to make sure it is working as expected, we need to manually create and check textures, so the existing test system does not work. and that is why we need unittest for pack/unpack (test-pack-unpack.ts).

but with unittest of pack/unpack, we can assume that pack and unpack should work as expected. based on this assumption, all other operator test should be able to fill into the operator test: given input tensor(s), expected output tensor(s) and attributes (if any), operator test should be able to handle all the rest.

there might be some missing features to run these operators in op test; but it should be better to implement the features in op test. there are already a lot of similar and duplicated code in each 'test-<some-op>.ts' now and these may cause burden for future maintenence.

Reply:
added depthToSpace to the op and node tests. Since the infra doesnt have a way to enable pack mode, we might need more changes to support running the packed matmul in op tests.
I agree the unit tests are due for some refactoring and consolidation, but I would still recommend adding unit tests for any upcoming features(ops, cache and etc). Unit tests are for testing specifically targeted code paths so that we can make sure the basic functionality works, that's why we have some manual and possibly duplicated lines in there to prefix the environments like texture cache, gl state, execution flow and etc. Also it's much easier to debug in the form of unit tests.